sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala

```diff
@@ -407,6 +407,34 @@ object PartitioningUtils {
       Literal(bigDecimal)
     }
 
+    val dateTry = Try {
+      // try and parse the date, if no exception occurs this is a candidate to be resolved as
+      // DateType
+      DateTimeUtils.getThreadLocalDateFormat.parse(raw)
+      // SPARK-23436: Casting the string to date may still return null if a bad Date is provided.
+      // This can happen since DateFormat.parse  may not use the entire text of the given string:
+      // so if there are extra-characters after the date, it returns correctly.
+      // We need to check that we can cast the raw string since we later can use Cast to get
+      // the partition values with the right DataType (see
+      // org.apache.spark.sql.execution.datasources.PartitioningAwareFileIndex.inferPartitioning)
+      val dateValue = Cast(Literal(raw), DateType).eval()
+      // Disallow DateType if the cast returned null
+      require(dateValue != null)
+      Literal.create(dateValue, DateType)
+    }
+
+    val timestampTry = Try {
+      val unescapedRaw = unescapePathName(raw)
+      // try and parse the date, if no exception occurs this is a candidate to be resolved as
+      // TimestampType
+      DateTimeUtils.getThreadLocalTimestampFormat(timeZone).parse(unescapedRaw)
```

Comment:
Can we save the parsing here? I think to cast string to `TimestampType` will also check if it can be parsed as timestamp?

Reply:
I don't think so, because in the cast we tolerate various timestamp format which parse doesn't support (please check the comment to `DateTimeUtils.stringToTimestamp`). So I'd not consider safe to remove this and anyway it would/may introduce unintended behavior changes.

Reply:
Since this changes the behavior of `PartitioningUtils.parsePartitions`, doesn't it change the result of another path in `inferPartitioning`?

Reply:
sorry, I am not sure I got 100% your question, may you elaborate it a bit more please?

Reply:
`inferPartitioning` will use `PartitioningUtils.parsePartitions` to infer the partition spec if there is no `userPartitionSchema`. It is used by `DataSource.sourceSchema`. Seems this change makes the partition directory previously parsing-able now unable to parse. Will it change behavior of other code path?

Reply:
Yes, you are right. the only change introduced is that some values which were previously wrongly inferred as dates, now will be inferred as strings. Everything else works as before.